### PR TITLE
Add table property to specify the physical file system location

### PR DESCRIPTION
**Pull request by zhenxiao**
Add a "location" property that allows specifying the physical file system location. A config option to disallow using this (or require the user to have the admin role), as many administrators wouldn't want users to create tables in any random location.
@electrum @cberner

### CODE REVIEW

**Comment by electrum** [comment]
Setting a location is specific to the Hive connector, so it shouldn't be added to `ConnectorAccessControl`.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java [hunk 1]
```diff
@@ -77,7 +79,8 @@ public HiveTableProperties(TypeManager typeManager, HiveClientConfig config)
                                 .map(name -> ((String) name).toLowerCase(ENGLISH))
                                 .collect(Collectors.toList())),
                         value -> value),
-                integerSessionProperty(BUCKET_COUNT_PROPERTY, "Number of buckets", 0, false));
+                integerSessionProperty(BUCKET_COUNT_PROPERTY, "Number of buckets", 0, false),
+                stringSessionProperty(LOCATION, "physical file system location", config.getLocation(), false));
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: Use `null` for the default. If it's not specified, we should use the location from the Hive database object.

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: Let's make the description `"File system location URI"`

**Comment by electrum** [comment]
I'm thinking we should make the config property `allow-set-location` to have three values: `true`, `admin-only`, `false`

**Comment by zhenxiao** [comment]
thank you, @electrum do you think we should have a `checkCanSetTableLocation` in `SqlStandardAccessControl` , and do the access control in CreateTableTask? In that case, we need to add `checkCanSetTableLocation` in `ConnectorAccessControl`
Or do you think we could just do the checking in HiveMetadata?

**Comment by electrum** [comment]
This is a Hive specific security check, so it should only be in the Hive connector. There is the complication that we need to do the check differently depending on `hive.security` setting.

We can add a `HiveAccessControl` interface that extends `ConnectorAccessControl`, then change Hive to use that interface everywhere. The method `checkCanSetTableLocation` can be added to the new Hive interface.

**Comment by electrum** [comment]
As a precursor to this change, I think it would help to cleanup the config handling for the Hive access control implementations:

1. Create new config classes for each of the access control implementations, i.e. `NoAccessControlConfig`, `SqlStandardAccessControlConfig`
2. Move the properties like `hive.allow-add-column` from `HiveClientConfig` to the new config classes
3. Update the implementations to take their own config class rather than `HiveClientConfig`
4. Bind the new configs in the respective modules, i.e. `NoSecurityModule` and `SqlStandardSecurityModule`

This will allow having a separate `hive.allow-set-location` for each module. For `NoAccessControl`, it should be a boolean (since there's no admin), but for `SqlStandardAccessControl` it can have the three states as mentioned above.

**Comment by zhenxiao** [comment]
thank you so much @electrum I will follow your idea
one more question, do you think we should pass `presto.security.AccessControl` to each connector, and do the access check inside each connector?
E.g. currently CreateTableTask is doing access control. And when it is checking, it is `presto.security.AccessControl`, no specific connector information available. So in CreateTableTask, there is no way to do `checkCanSetTableLocation`. Do you think we need to pass `presto.security.AccessControl` to Hive connector, and do the check inside HiveMetadata? Or did I misunderstanding something?

**Comment by electrum** [comment]
We should have HiveMetadata do the check by calling canSetTableLocation.

**Pull request closed without merging**